### ISSUE TITLE
Fatal Python error: Floating point exception in tracker's first real align() call (TUM freiburg1_desk, reproducible in pure vanilla source)

### ISSUE THREAD

**Issue by bjoernellens1** [Lab-of-AI-and-Robotics/GS_ICP_SLAM#61]
## Summary

Floating point exception crashes the tracker process inside vendored `fast_gicp` 's native `pygicp.FastGICP.align()` call, on the very first real per-frame tracking call (frame index 1) of the TUM `freiburg1_desk` RGB-D sequence. 

## Environment

- `GS_ICP_SLAM` commit `5f996a8` ("Update README.MD")
- Vendored `submodules/fast_gicp` at its own `gs_icp_slam` branch tip, commit `e2954b1` ("Tackle error : NaN mahalanobis") — no later commits exist on that branch
- Base image: `pytorch/pytorch:2.9.1-cuda12.8-cudnn9-devel` (CUDA 12.8, Python 3.11)
- Real NVIDIA A100-40GB GPU (k8s cluster, CUDA MPS enabled for multi-process GPU sharing — unrelated to the crash itself, confirmed by reproducing on a dedicated/non-shared GPU too)
- CPU FastGICP path (PCL KdTree + OpenMP), not the CUDA `fast_vgicp` voxel path — the tracker never calls a CUDA-accelerated registration method

## Steps to reproduce

1. Run GS-ICP-SLAM's RGB-D tracker (`gs_icp_slam_unlimit.py` / equivalently `mp_Tracker_unlimit.py`'s `Tracker.tracking()`) on the TUM `freiburg1_desk` sequence (`rgbd_dataset_freiburg1_desk`).
2. The tracker's `get_images()` bulk-loads all RGB-D frames successfully (harmless, `cv2.imread`/`o3d.io.read_image`, unrelated to GICP).
3. The real per-frame tracking loop begins. On the very first real alignment call (`self.reg.align(initial_pose)` at `mp_Tracker_unlimit.py:173`, frame index 1 — frame 0 is a separate initialization branch), the process crashes with:

```
Fatal Python error: Floating point exception

Current thread (most recent call first):
  File "mp_Tracker_unlimit.py", line 173 in tracking
  File "mp_Tracker_unlimit.py", line 98 in run
  File "gs_icp_slam_unlimit.py", line 114 in tracking
  File "multiprocessing/process.py", line 108 in run
  ...
```

(Traceback captured via Python's `faulthandler.enable()`, added purely as a diagnostic — it does not change any behavior.)

Since this is a hardware floating-point trap (SIGFPE), not a catchable Python exception, the process dies immediately and unrecoverably at the OS level.

## What we ruled out (real, live-tested, all unsuccessful)

We spent significant effort chasing this and want to save others the trouble by sharing exactly what we tried and ruled out, each independently built and tested on a real GPU:

1. Widening `fast_gicp_impl.hpp`'s `NORMALIZED_ELLIPSE` covariance-regularization exact-zero guard (`svd.singularValues()(1) == 0`) to an epsilon floor (`1e-3`) — still crashed.
2. Forcing `FastGICP`'s constructor default away from `RegularizationMethod::NORMALIZED_ELLIPSE` to `MIN_EIG` — still crashed.
3. Loosening `max_correspondence_distance` from `0.02` to `0.2` (to rule out a zero-correspondence degenerate case) — still crashed.
4. Guarding `LsqRegistration::step_lm()`'s Levenberg-Marquardt `rho` division (and the initial `lm_lambda_` computation) against non-finite/near-zero denominators — still crashed.
5. Disabling `FE_DIVBYZERO`/`FE_INVALID`/`FE_OVERFLOW` hardware FP traps (`feclearexcept`/`fedisableexcept`) at the pygicp Python binding boundary, immediately before every `reg.align()` call — still crashed (this strongly suggests the trap may be occurring somewhere the host-side `fedisableexcept()` doesn't reach, e.g. deep in a vectorized/SIMD code path, or it isn't actually a classic divide-by-zero trap at all).
6. Forcing `OMP_NUM_THREADS=1` to rule out an OpenMP thread-scheduling race — still crashed identically (this was initially misread as a fix due to conflating `get_images()`'s progress bar with the real tracking loop's progress bar — both are `tqdm(total=592)`, which look identical in log output. Once we distinguished them via `faulthandler`, the crash was still present and identical regardless of thread count.)
7. Widening the `mahalanobis_[i] = RCR.inverse()` guard (from commit `e2954b1`, "Tackle error : NaN mahalanobis") from an exact-zero determinant check (`if (RCR.determinant() == 0)`) to an epsilon guard (`std::abs(RCR.determinant()) < 1e-9`), falling back to the safe pseudo-inverse in more cases — still crashed identically.

We also confirmed this is **not** a race condition: 3/3 identical crashes across multiple independent retry-from-scratch attempts, at the same file/line, every time.

## Why we're filing this

We searched this repo's issues and the `fast_gicp` fork's issues (and upstream `koide3/fast_gicp`) and found no exact match. The closest related reports are:
- #34 in this repo ("race condition?") — a real, separate, unresolved tracker/mapper shared-memory race in `mp_Mapper.py`.
- #37 in the `fast_gicp` fork — reports `align()` hanging + a missing `set_max_knn_distance` attribute on this same `gs_icp_slam` branch, suggesting other known instability in this fork's `align()` path, but not this specific SIGFPE.

We don't have a fix, but we have a precise, reproducible repro and a list of ruled-out hypotheses that should save whoever picks this up real time. Happy to provide more detail (full build logs, exact patch diffs we tried) if useful.

**Comment by rjaras** [comment]
I ran into the same crash with the Replica dataset and I believe the root cause is the `fast_gicp` submodule rather than CUDA/Python/PyTorch.

In particular, the `gs_icp_slam` branch at commit `e2954b1` appears to have a numerical issue with **PCL 1.12** (the default version installed by `libpcl-dev` on Ubuntu 22.04). This matches the reports in #25, where users confirmed that downgrading to **PCL 1.10** fixes the same crash.

I also tested this myself: **downgrading from PCL 1.12 to PCL 1.10 resolves the crash**.

This is also consistent with #30, where the author mentions testing on Ubuntu 20.04, which ships PCL 1.10. The specific filtered target/covariance path used by GS-ICP-SLAM (`set_target_filter` / `calculate_target_covariance_with_filter`) also does not appear to be covered by `fast_gicp`'s own Python tests.

**Comment by xijin369** [comment]
# SIGFPE in `pygicp.FastGICP.align()` — root-cause + minimal patch (GS_ICP_SLAM `5f996a8`, fast_gicp `e2954b1`)

**Summary.** The tracker process dies with `SIGFPE` on the **first** frame, inside
`self.reg.align(initial_pose)` (`mp_Tracker.py:199`). The signal is raised *inside
`pcl::transformPointCloud()`*, called from `fast_gicp::LsqRegistration::computeTransformation()`
(`lsq_registration_impl.hpp:76-77`) — the point cloud and the transformation are both verified
finite, so it is not a NaN/Inf input problem. Replacing that single PCL call with an equivalent
explicit affine loop makes the tracker run. I also found and fixed a second, independent
integer-divide-by-zero (empty kNN set → `Eigen ... rowwise().mean()` on a `4x0` matrix) in the
covariance estimation. Both changes are header-only; patch attached.

## 1. Reproduction

```bash
python gs_icp_slam.py --dataset_path dataset/TUM/rgbd_dataset_freiburg1_desk \
  --config configs/TUM/rgbd_dataset_freiburg1_desk.txt --output_path output/fr1_desk \
  --keyframe_th 0.81 --knn_maxd 99999.0 --overlapped_th 1e-3 \
  --max_correspondence_distance 0.03 --trackable_opacity_th 0.09 \
  --overlapped_th2 1e-3 --downsample_rate 5 --save_results
```

Symptom: `Floating point exception (core dumped)` — Python cannot catch it, and the second
process (`Mapper`) keeps spinning, so it looks like a hang / "stuck at 0%".
Reproducible on `rgbd_dataset_freiburg1_desk` and `rgbd_dataset_freiburg1_xyz`.

## 2. Evidence

1. `faulthandler` → the process stops exactly inside the `pygicp` call.
2. Core dump (`/proc/sys/kernel/core_pattern=/tmp/core.%p`) opened with gdb:

```
Program terminated with signal SIGFPE, Arithmetic exception.
#0  void pcl::transformPointCloud<pcl::PointXYZ, float>(pcl::PointCloud<pcl::PointXYZ> const&,
        pcl::PointCloud<pcl::PointXYZ>&, Eigen::Matrix<float,4,4,0,4,4> const&, bool)
        () from .../libfast_gicp.so
#1  fast_gicp::LsqRegistration<pcl::PointXYZ, pcl::PointXYZ>::computeTransformation(
        pcl::PointCloud<pcl::PointXYZ>&, Eigen::Matrix<float,4,4,0,4,4> const&)  () from .../libfast_gicp.so
#2  ?? () from .../pygicp.cpython-310-x86_64-linux-gnu.so
#3  ?? () from .../pygicp.cpython-310-x86_64-linux-gnu.so
#4  ?? () from .../pygicp.cpython-310-x86_64-linux-gnu.so
```

3. A diagnostic inserted right before the failing call shows `nonfinite_points=0` and
   `transform_allFinite=1` for every `align()` that dies → the input data is fine.

## 3. Fix

**(a) `include/fast_gicp/gicp/impl/lsq_registration_impl.hpp`** — `LsqRegistration::computeTransformation()`
(repo lines 76-77):

```cpp
final_transformation_ = x0.cast<float>().matrix();
pcl::transformPointCloud(*input_, output, final_transformation_);   // <-- SIGFPE raised inside
```

replaced by an explicit affine transform (same semantics, no PCL SIMD path). **This alone stops the crash.**

**(b) `include/fast_gicp/gicp/impl/fast_gicp_impl.hpp`** — in the four
`calculate_{source,target}_covariances_with_filter()` overloads: if a point has no neighbour with
`k_sq_distances[j] < knn_max_distance_`, then `num_reliable_neighbors == 0` and the following
`Eigen::Matrix<double,4,-1> neighbors(4, 0)` makes `neighbors.rowwise().mean()` (and the subsequent
covariance math) divide by zero columns. Added:

```cpp
if (num_reliable_neighbors == 0) {
  continue;   // avoid Eigen rowwise().mean() on a 4x0 matrix (integer division by zero -> SIGFPE)
}
```

Note on ordering: (b) was found first while bisecting the crash and is **not sufficient alone** on
TUM (the crash persists), but it is a genuine divide-by-zero that fires on sparse/empty
neighbourhoods (e.g. depth maps with holes — my own structured-light captures at 640×480), so both
are worth keeping.

## 4. Verification (after the patch, `--download`-free local TUM data)

`rgbd_dataset_freiburg1_desk`, 592 associated frames, the command from §1 (two unseeded draws —
see §5 for the run-to-run spread):

| run | wall time | System FPS | ATE RMSE | PSNR | SSIM | LPIPS |
|---|---|---|---|---|---|---|
| 1 | 60 s, exit=0 | 29.97 | 2.95 cm | 17.18 | 0.695 | 0.325 |
| 2 | 47 s, exit=0 | 29.94 | 2.99 cm | 17.16 | 0.698 | 0.319 |

The same patched build also runs to completion on my own RGB-D captures
(ORBBEC Astra Pro, 640×480, `DEPTH_1_MM` depth stored as uint16 mm PNG, `depth_scale=1000`,
3060 frames): 5.5M gaussians, no SIGFPE, `System FPS 20.99`.
So the fix is not specific to the TUM sequences.

## 5. Run-to-run variance of the metrics (not caused by this patch)

While bisecting I noticed fr1/desk is **not reproducible run-to-run** on the patched build
(same command, same data, same build):

| setup | ATE RMSE (cm) | PSNR |
|---|---|---|
| unseeded, 5 runs | 2.52 / 2.95 / 2.99 / **33.20** / **103.52** | 14.23 – 17.18 |
| seeds fixed, 3 runs | 2.39 / 3.06 / 2.68 | 17.11 – 17.29 |

Cause is in the Python side, not in the patch: `mp_Mapper.py:205` picks the training view with an
*unseeded* `random.choice(...)`, `scene/dataset_readers.py:85` uses an *unseeded*
`np.random.choice(...)`, and the tracker/mapper are two `spawn`-ed processes
(`gs_icp_slam.py:66`), each with its own RNG state. Fixing the seeds inside `Tracker.run()` /
`Mapper.run()` removes the outliers (and leaves a ~±0.3 cm jitter from the async interleaving).

So a single run's ATE is not a meaningful pass/fail criterion here; the criterion I used is
**"no SIGFPE + the tracker completes the whole sequence + PSNR/SSIM in the usual range"**.
(I initially mis-read one bad draw as a regression caused by my own patch cleanup — it was not.)

## 6. Environment

- `GS_ICP_SLAM` `5f996a8`, `submodules/fast_gicp` `e2954b1`
  (https://github.com/Lab-of-AI-and-Robotics/fast_gicp.git)
- Ubuntu 22.04 (WSL2), gcc 11.4, **PCL 1.12.1**, **Eigen 3.4.0**, **pybind11 2.13.6**
  (bumped from the vendored 2.6.3, which does not support Python 3.10), CUDA 12.8, torch 2.8.0+cu128,
  Python 3.10, GPU RTX 5060 Ti (sm_120)
- Rebuilt in-place with `pip install -e . --no-build-isolation` inside `submodules/fast_gicp`
  (patch is header-only, no CMake change needed)


<details>
<summary><b>Full patch</b> (2 files, header-only) — also attached as <code>gsicp-sigfpe-fix.patch</code></summary>

```diff
diff --git a/include/fast_gicp/gicp/impl/fast_gicp_impl.hpp b/include/fast_gicp/gicp/impl/fast_gicp_impl.hpp
index b7aa14a..fa2f606 100644
--- a/include/fast_gicp/gicp/impl/fast_gicp_impl.hpp
+++ b/include/fast_gicp/gicp/impl/fast_gicp_impl.hpp
@@ -407,6 +407,11 @@ bool FastGICP<PointSource, PointTarget, SearchMethodSource, SearchMethodTarget>:
       }
     }
 
+    if (num_reliable_neighbors == 0) {
+      // no neighbor within knn_max_distance_: skip, else rowwise().mean() below divides by cols()==0
+      continue;
+    }
+
     Eigen::Matrix<double, 4, -1> neighbors(4, num_reliable_neighbors);
     for (int j = 0; j < k_indices.size(); j++) {
       if (k_sq_distances[j] < knn_max_distance_){
@@ -509,6 +514,11 @@ bool FastGICP<PointSource, PointTarget, SearchMethodSource, SearchMethodTarget>:
       }
     }
 
+    if (num_reliable_neighbors == 0) {
+      // no neighbor within knn_max_distance_: skip, else rowwise().mean() below divides by cols()==0
+      continue;
+    }
+
     Eigen::Matrix<double, 4, -1> neighbors(4, num_reliable_neighbors);
     for (int j = 0; j < k_indices.size(); j++) {
       if (k_sq_distances[j] < knn_max_distance_){
@@ -622,6 +632,11 @@ bool FastGICP<PointSource, PointTarget, SearchMethodSource, SearchMethodTarget>:
       }
     }
 
+    if (num_reliable_neighbors == 0) {
+      // no neighbor within knn_max_distance_: skip, else rowwise().mean() below divides by cols()==0
+      continue;
+    }
+
 
     Eigen::Matrix<double, 4, -1> neighbors(4, num_reliable_neighbors);
     for (int j = 0; j < k_indices.size(); j++) {
@@ -743,6 +758,11 @@ bool FastGICP<PointSource, PointTarget, SearchMethodSource, SearchMethodTarget>:
       }
     }
 
+    if (num_reliable_neighbors == 0) {
+      // no neighbor within knn_max_distance_: skip, else rowwise().mean() below divides by cols()==0
+      continue;
+    }
+
     Eigen::Matrix<double, 4, -1> neighbors(4, num_reliable_neighbors);
     for (int j = 0; j < k_indices.size(); j++) {
       if (k_sq_distances[j] < knn_max_distance_){
diff --git a/include/fast_gicp/gicp/impl/lsq_registration_impl.hpp b/include/fast_gicp/gicp/impl/lsq_registration_impl.hpp
index 5810efb..80352b5 100644
--- a/include/fast_gicp/gicp/impl/lsq_registration_impl.hpp
+++ b/include/fast_gicp/gicp/impl/lsq_registration_impl.hpp
@@ -74,7 +74,22 @@ void LsqRegistration<PointTarget, PointSource>::computeTransformation(PointCloud
   }
 
   final_transformation_ = x0.cast<float>().matrix();
-  pcl::transformPointCloud(*input_, output, final_transformation_);
+
+  {  // explicit affine transform instead of pcl::transformPointCloud() [SIGFPE in core dump]
+    output.points.resize(input_->points.size());
+    const Eigen::Matrix4f& tf = final_transformation_;
+    for (std::size_t i = 0; i < input_->points.size(); i++) {
+      const PointSource s = input_->points[i];
+      PointSource d = s;
+      d.x = tf(0, 0) * s.x + tf(0, 1) * s.y + tf(0, 2) * s.z + tf(0, 3);
+      d.y = tf(1, 0) * s.x + tf(1, 1) * s.y + tf(1, 2) * s.z + tf(1, 3);
+      d.z = tf(2, 0) * s.x + tf(2, 1) * s.y + tf(2, 2) * s.z + tf(2, 3);
+      output.points[i] = d;
+    }
+    output.width = input_->width;
+    output.height = input_->height;
+    output.is_dense = input_->is_dense;
+  }
 }
 
 template <typename PointTarget, typename PointSource>
```

</details>


[gdb-core-backtrace.txt](https://github.com/user-attachments/files/32507088/gdb-core-backtrace.txt)
[gsicp-sigfpe-fix.patch](https://github.com/user-attachments/files/32507087/gsicp-sigfpe-fix.patch)
